src/Database/Redis/ProtocolPipelining.hs

```diff
@@ -28,29 +28,43 @@ import           Control.Exception
 import           Control.Monad
 import qualified Scanner
 import qualified Data.ByteString as S
+import qualified Data.ByteString.Lazy as L
 import           Data.IORef
 import           Data.Typeable
 import           Network
 import qualified Network.BSD as BSD
 import qualified Network.Socket as NS
+import qualified Network.TLS as TLS
 import           System.IO
 import           System.IO.Error
 import           System.IO.Unsafe
 
 import           Database.Redis.Protocol
 
 
-data Connection = Conn
-  { connHandle     :: Handle        -- ^ Connection socket-handle.
-  , connReplies    :: IORef [Reply] -- ^ Reply thunks for unsent requests.
-  , connPending    :: IORef [Reply]
-    -- ^ Reply thunks for requests "in the pipeline". Refers to the same list as
-    --   'connReplies', but can have an offset.
-  , connPendingCnt :: IORef Int
-    -- ^ Number of pending replies and thus the difference length between
-    --   'connReplies' and 'connPending'.
-    --   length connPending  - pendingCount = length connReplies
-  }
+data Connection
+  = Conn
+    { connHandle     :: Handle        -- ^ Connection socket-handle.
+    , connReplies    :: IORef [Reply] -- ^ Reply thunks for unsent requests.
+    , connPending    :: IORef [Reply]
+      -- ^ Reply thunks for requests "in the pipeline". Refers to the same list as
+      --   'connReplies', but can have an offset.
+    , connPendingCnt :: IORef Int
+      -- ^ Number of pending replies and thus the difference length between
+      --   'connReplies' and 'connPending'.
+      --   length connPending  - pendingCount = length connReplies
+    }
+  | TLSConn
```

Comment:
Thanks for the work! Few things:
- please avoid mixing constructors and field accessors, this code allows doing `tlsConnContext` on `Conn`-constructed data-structure, leading to a possible runtime exception, I also don't like the duplication of the replies/pending/pendingcnt fields. My proposal is: make a datatype `data ConnHandleOrTls = ConnHandle Handle | ConnTlsContext TLS.Context`, but I'm open to alternatives
- would be nice to avoid the reformatting as much as possible (not critical in a PR of this size)